### PR TITLE
Add mmdarwin module

### PR DESCRIPTION
**Pull request by m4jr0**
Add a mmdarwin module, allowing to use the [Darwin cyber security framework](https://github.com/VultureProject/darwin) (which will be available soon) from Rsyslog.

### CODE REVIEW

**Comment by rgerhards** [comment]
thx for the PR, sounds great!

Question now: is there any way we can compile it inside CI? Can we add some tests?

**Comment by rgerhards** [comment]
@m4jr0 did you see the CI failures?

File: contrib/mmdarwin/mmdarwin.c [hunk 1]
```diff
@@ -407,16 +407,19 @@ CODESTARTnewActInst
 		}  else if (!strcmp(actpblk.descr[i].name, "response")) {
 			char *response = es_str2cstr(pvals[i].val.d.estr, NULL);
 
-			if (!strcmp(response, "no") != 0) {
+			if (!(strcmp(response, "no") != 0)) {
```
Comment (rgerhards) [inline contrib/mmdarwin/mmdarwin.c]: I think these boolean expressions need clarification.

I suggest to replace them by either
`strcmp(resonse, "no") == 0`
or the well-knwon
`!strcmp(resonse, "no")` - I think this should have been it in the first place,.

Comment (m4jr0) [reply contrib/mmdarwin/mmdarwin.c]: Thank you for improving the readability of the code.

**Comment by m4jr0** [comment]
Sorry for replying so late. I've made some adjustments in the code according to the CI failures.

**Comment by rgerhards** [comment]
Can we add some tests?

**Comment by rgerhards** [comment]
Question: Is GPLv3 a hard requirement? We are trying to move rsyslog to ASL 2.0. For a contributed module, GPLv3 surely is no problem, but at some point in future we may delete it from our tree because we want to go pure ASL 2.0.

**Comment by rgerhards** [comment]
@m4jr0 The copyright statement is not fully correct. I definitely see parts of my code inside it. Please let me know which module this is based on so that I can apply the proper additional copyrights.

**Comment by rgerhards** [comment]
@m4jr0 please add doc to https://github.com/rsyslog/rsyslog-doc/

**Comment by rgerhards** [comment]
I have amended the PR with some basic dynamic tests and also fixed an issue that the test detected. LGTM.com is now also activated.

**Comment by rgerhards** [comment]
@m4jr0 the merge window for 8.1904.0 closes friday. Do you think you can do the remaining work until friday morning?

**Comment by m4jr0** [comment]
Hello,

GPLv3 is not a hard requirement, so I've changed the copyright statements to ASL 2.0. By the way, this module has been inspired by the imtcp and mmdblookup modules.

I've created a pull request on rsyslog-doc [here](https://github.com/rsyslog/rsyslog-doc/pull/804) that includes some documentation written for the module.

Should I add more tests? I'm not sure if everything will be done by Friday morning if that's the case.

**Comment by davidelang** [comment]
looking at this (starting from the documentation) , it would be better if you could pass a json stanza instead of a position dependent array.

Is this something where you expect to get an answer back instantly? any delays in getting a response can stall all log processing. You may want to think in terms of dispatching the log to darwin and have it's action be to generate a new message into rsyslog rather than making a call to darwin for each message.

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging 6b71b2d772cb162071379a63fa86d9a8202660f2 into fcb83fb247e70d53c63f2c34c9b5339c9310ef80 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-bc7d65d3ac9c23969f53abd6c085797ad14e4ef5)

**new alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
> Should I add more tests? I'm not sure if everything will be done by Friday morning if that's the case.

I would say it would be extremely good to have the tests. We have often seen code fail on platforms that were not used during development. So I would prefer to merge a release later when I can be sure we have tests and have run them before merging.

**Comment by rgerhards** [comment]
removed target milestone as I do not see any progress on this PR

**Comment by m4jr0** [comment]
Hi,

@davidelang Currently, a Darwin filter expects some parameters, which are set in a defined order. So I don't understand why it would be better to use JSON instead? Maybe I misunderstood your point, or is it the mmdarwin documentation that is not clear enough?

Also, Darwin is expected to answer back instantly. Are you suggesting to make it able to generate the enriched log lines itself instead of mmdarwin?

@rgerhards I've added some tests to mmdarwin, which require a Darwin filter to run at the same time. Since it has not been released on Github yet (which is beyond my control, but it should be out in a few days/weeks), is that a problem for you? What kind of tests should I add besides the ones listed below?

Currently, the tests for mmdarwin are:
 - invalid socket provided (your contribution)
 - invalid socket provided (with Valgrind checks) (your contribution)
 - no parameters provided (your contribution)
 - valid parameters provided, except the for fields (which are malformed)
 - valid parameters provided, except the for fields (which are malformed) (with Valgrind checks)
 - valid parameters provided, but the field given does not exist in the JSON log line
 - valid parameters provided, but the field given does not exist in the JSON log line (with Valgrind checks)
 - valid parameters provided, but the fields given is an empty array
 - valid parameters provided, but the fields given is an empty array (with Valgrind checks)
 - valid parameters provided, but the key given is empty
 - valid parameters provided, but the key given is empty (with Valgrind checks)
 - valid parameters provided, expecting everything to work correctly
 - valid parameters provided, expecting everything to work correctly (with Valgrind checks)

PS: the commented-out code alert points to a quick C example snippet that can be compiled to generate a filter code. Would you prefer to have it removed?

**Comment by rgerhards** [comment]
any interest anyone in completing this work?

**Comment by m4jr0** [comment]
@rgerhards What kind of test should I add besides the ones I added before? What can I do to improve the code coverage?

**Comment by m4jr0** [comment]
I wasn't sure to fix the LGTM issue or not, because it only refers to a code snippet that has been commented out on purpose to show how to generate a filter code for Darwin. But I'll remove the snippet from protocol.h and put it in the documentation instead.

Some changes has been done to mmdarwin. I'll update the code today and see what I can do to improve the code coverage. Thanks.

**Comment by rgerhards** [comment]
> I wasn't sure to fix the LGTM issue or not, because it only refers to a code snippet that has been commented out on purpose to show how to generate a filter code for Darwin. But I'll remove the snippet from protocol.h and put it in the documentation instead.

That's a great plan. An alternative to the doc is a sample c file, just FYI. Inside the doc it's probably more visible.

Commented out code is really eval. It's hard to remove once there, so we do not accept any new ones into the codebase and work on removing the few leftovers we have.
 
> Some changes has been done to mmdarwin. I'll update the code today and see what I can do to improve the code coverage. Thanks.

I'll keep an eye on the PR and am ready to lend a helping hand if needed. Would really love to get this into master. Thanks for following up!

**Comment by rgerhards** [comment]
I notice that the CodeCov UI pretty well hides the statement coverage for the PR. Here is a direct link: https://codecov.io/gh/rsyslog/rsyslog/pull/3556/diff

File: contrib/mmdarwin/mmdarwin.c [hunk 2]
```diff
@@ -1,17 +1,10 @@
...
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
+/// \file     mmdarwin.c
+/// \authors  gcatto
+/// \version  1.0
+/// \date     08/11/18
+/// \license  GPLv3
```
Comment (rgerhards) [inline contrib/mmdarwin/mmdarwin.c]: Is GPLv3 a hard requirement? rsyslog is (slowly but steadily) moving towards ASL 2.0 and being GPLv3 will prevent this module from ever become an official one.

Comment (m4jr0) [reply contrib/mmdarwin/mmdarwin.c]: My mistake!

Comment (rgerhards) [reply contrib/mmdarwin/mmdarwin.c]: thx a lot!

**Comment by m4jr0** [comment]
I've made the changes to mmdarwin. I'm not sure about the **buildbot/rsyslog freebsd rsyslog** check though.

**Comment by m4jr0** [comment]
@rgerhards I'm not sure about the **[buildbot/rsyslog freebsd rsyslog](https://build.rsyslog.com/#/builders/63/builds/2472/steps/5/logs/stdio)** build. Do you have any idea what made it fail?

```
make  ourtail  tcpflood chkseq  msleep randomgen  diagtalker uxsockrcvr  syslog_caller  inputfilegen  minitcpsrv  omrelp_dflt_port  mangle_qi  have_relpSrvSetOversizeMode  have_relpEngineSetTLSLibByName  test_id   
  CC       ourtail.o
  CCLD     ourtail
  CC       tcpflood-tcpflood.o
tcpflood.c: In function 'sendMessages':
tcpflood.c:684:4: warning: passing argument 5 of 'sendto' from incompatible pointer type [enabled by default]
    lenSend = sendto(udpsock, buf, lenBuf, 0, &udpRcvr, sizeof(udpRcvr));
    ^
In file included from tcpflood.c:103:0:
/usr/include/sys/socket.h:612:9: note: expected 'const struct sockaddr *' but argument is of type 'struct sockaddr_in *'
 ssize_t sendto(int, const void *,
         ^
  CCLD     tcpflood
```

**Comment by rgerhards** [comment]
> @rgerhards I'm not sure about the **[buildbot/rsyslog freebsd rsyslog](https://build.rsyslog.com/#/builders/63/builds/2472/steps/5/logs/stdio)** build. Do you have any idea what made it fail?

This sounds strange.

**Comment by rgerhards** [comment]
@m4jr0 ok, that was not the real error, just a compile nit. The error was later below - and it was a false positives. Unfortunately we have them from time to time... I've re-initiated the CI run and expect it to run fine. Stand by, I'll let you know the end result.

**Comment by m4jr0** [comment]
@rgerhards How should I proceed with the last failed build? Should I squash the commits? I'm not sure what to do because of the merged commits in between mines.

**Comment by rgerhards** [comment]
sorry for the close/re-open - browser too slow, accidently hit "close". If CI now has another flake, pls ignore.

To your question: ultimately, this adds a feature, so we need a 1-commit PR. I would prefer if you can squash/rebase. If not, I'll do so as part of the final merge. But it's easier IMHO if it is done earlier. But use whatever you prefer.

No problems with my commits - just squash them without mentioning me. It was really nothing of substance (and all being ASL 2.0 also makes us clean). 

I think commit was ready for a new manual review. Not sure if I mange to do it today. Probably Friday (tomorrow is a public holiday at my location). My first impression is that we are close to merging. My main focus would be the CodeCov. I think there are some codepaths which should be covered by test (send/receive) - but I am not sure if this can be done without Darwin or a mock-up.

**Comment by m4jr0** [comment]
I meant, because of all the merges from master that are mixed up with our commits, I'm not sure what to do with them: I'm not used to call `git rebase -i`. What should I do with them?

**Comment by rgerhards** [comment]
I need to have a look. Looks like a mess due to the missing rebases.

**Comment by rgerhards** [comment]
@m4jr0 pls let me know when you think the PR is ready. I'll then review. Thx! :-)

**Comment by m4jr0** [comment]
> @m4jr0 pls let me know when you think the PR is ready. I'll then review. Thx! :-)

I think the PR is ready. :)

**Comment by rgerhards** [comment]
thx again for the PR and your persistence in working through to the final version. Much appreciated!

**Pull request merged**